kusto/ingest/status.go

```diff
@@ -0,0 +1,291 @@
+	// When <see cref="IngestionReportMethod"/> is set to 'Table', call <see cref="IKustoIngestionResult.GetIngestionStatusBySourceId"/> or
+	// <see cref="IKustoIngestionResult.GetIngestionStatusCollection"/> to retrieve the most recent ingestion status.
+	Status StatusCode
+
+	// IngestionSourceID - A unique identifier representing the ingested source. It can be supplied during the ingestion execution.
+	IngestionSourceID uuid.UUID
+
+	// The URI of the blob, potentially including the secret needed to access
+	// the blob. This can be a filesystem URI (on-premises deployments only),
+	// or an Azure Blob Storage URI (including a SAS key or a semicolon followed
+	// by the account key).
+	IngestionSourcePath string
+
+	// Database - The name of the database holding the target table.
+	Database string
+
+	// Table - The name of the target table into which the data will be ingested.
+	Table string
+
+	// UpdatedOn - The last updated time of the ingestion status.
+	UpdatedOn time.Time
+
+	// OperationID - The ingestion's operation ID.
+	OperationID uuid.UUID
+
+	// ActivityID - The ingestion's activity ID.
+	ActivityID uuid.UUID
+
+	// ErrorCode In case of a failure, indicates the failure's error code.
+	// TODO [Yochai, July 2020] make this into a const list
+	ErrorCode string
+
+	// FailureStatus - In case of a failure, indicates the failure's status.
+	FailureStatus FailureStatusCode
+
+	// Details - In case of a failure, indicates the failure's details.
+	Details string
+
+	// OriginatesFromUpdatePolicy - In case of a failure, indicates whether or not the failure originated from an Update Policy.
+	OriginatesFromUpdatePolicy bool
+}
+
+// newStatusRecord creates a new record initialized with defaults and user provided data.
+func newStatusRecord() StatusRecord {
+	rec := StatusRecord{
+		Status:                     Failed,
+		IngestionSourceID:          uuid.Nil,
+		IngestionSourcePath:        "Undefined",
+		Database:                   "Undefined",
+		Table:                      "Undefined",
+		UpdatedOn:                  time.Now(),
+		OperationID:                uuid.Nil,
+		ActivityID:                 uuid.Nil,
+		ErrorCode:                  "Unknown",
+		FailureStatus:              Unknown,
+		Details:                    "",
+		OriginatesFromUpdatePolicy: false,
+	}
+
+	return rec
+}
+
+// FromProps takes in data from ingestion options.
+func (r *StatusRecord) FromProps(props properties.All) {
+	r.IngestionSourceID = props.Source.ID
+	r.Database = props.Ingestion.DatabaseName
+	r.Table = props.Ingestion.TableName
+	r.UpdatedOn = time.Now()
+
+	if props.Ingestion.BlobPath != "" && r.IngestionSourcePath == "Undefined" {
+		r.IngestionSourcePath = props.Ingestion.BlobPath
+	}
+}
+
+// FromMap converts an ingestion status record to a key value map.
+func (r *StatusRecord) FromMap(data map[string]interface{}) {

```

Comment:
So, this CL seems incomplete, which I imagine means you checked in some code already.  Though it could be I don't do a lot of github reviews.

I don't see GetEntityReference here, which would probably confirm my suspicions.  But from what I'm seeing, FromMap() is probably converting a json.Unmarshal() into a map[string]interface{}.  

I don't think we should probably be doing that, instead we should be doing a json.Unmarshal directly into a StatusRecord.  

json decoding not being anyone's favorite, my guess is that there were some problems and that lead to this conversion into map[string]interface{} and then into the concrete type.

If that isn't wrong, then we should talk about what the problems were and eliminate all this map[string]interface{} code.

Reply:
Oh, ......

Well that is just horrible.  As a note, if you ever see map[string]interface{}, it is almost always someone being lazy about a JSON conversion.  

This is absolutely terrible that they are giving this back to you instead of a struct.

So ignore my comment.